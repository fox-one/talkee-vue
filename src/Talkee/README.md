# Talkee

## Example

```vue
<template>
  <div>
    <h2 class="title">{{ title }}</h2>
    <div>帖子内容</div>
    <talkee
      :siteId="1"
      :slug="1234"
      apiBase="https://talkee-test-api.firesbox.com/api"
      loginUrl="https://mixin-oauth.firesbox.com/?client_id=65fdd5af-5d3f-496a-9d62-b5cdb4d8396f&scope=PROFILE:READ+PHONE:READ&redirect_url="
      @resize="handleResize"
      :renderOpts="{
        reply: false
      }"
    />
  </div>
</template>

<script>
  export default {
    data: {
      title: '帖子标题'
    },
    method: {
      handleResize: function () {
        console.info('resize!');
      }
    }
  }
</script>

<style lang="scss" scoped>
.title {
  font-size: 20px;
}
</style>
```
