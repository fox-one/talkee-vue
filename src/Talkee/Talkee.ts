import {
  Component,
  Vue,
  Prop,
  Watch
} from 'vue-property-decorator';
import TalkeeSDK from '@links-japan/talkee';
import classnames from '@utils/classnames';
/* import types */
import type { CreateElement, VNode } from 'vue';

export type TALKEE_INSTANCE = TalkeeSDK;

@Component
export class Talkee extends Vue {
  @Prop({ type: String, default: '' }) private apiBase!: string;
  @Prop({ type: String, default: '' }) private loginUrl!: string;
  @Prop({ type: String, default: 'talkee' }) private prefixCls!: string;
  @Prop({ type: Number, default: 1 }) private siteId!: number;
  @Prop({ type: [String, Number], default: void 0 }) private slug!: number | string;
  @Prop({ type: Boolean, default: true }) private expandable!: string;
  @Prop({ type: Array, default: () => [] }) private tweetTags!: string[];
  @Prop({ type: Object, default: () => { } }) private renderOpts!: {
    metabar?: boolean;
    subcomment?: boolean;
    tweet?: boolean;
    reply?: boolean;
  };

  private talkee: undefined | TALKEE_INSTANCE;

  private size = {
    width: -Infinity,
    height: -Infinity
  };

  private timer: undefined | ReturnType<typeof setTimeout>;

  public get classes() {
    return classnames(this.prefixCls);
  }

  private handleError(e: any) {
    this.$emit('error', e);
  }

  private resizeListener() {
    this.$nextTick(() => {
      const talkee = this.$refs.talkee as HTMLElement;
      const {
        width,
        height
      } = talkee.getBoundingClientRect();
      const w = talkee.offsetWidth ?? width;
      const h = talkee.offsetHeight ?? height;

      if (this.size.width !== w || this.size.height !== h) {
        this.size.width = w;
        this.size.height = h;
        this.$emit('resize', this.talkee);
      }

      this.timer = setTimeout(() => {
        this.resizeListener();
      }, 1500);
    });
  }

  private initTalkee() {
    if (this.talkee) return;
    if (!this.siteId || !this.slug || !this.apiBase || !this.loginUrl) {
      console.error(`
        siteId: ${this.siteId},
        slug: ${this.slug},
        apiBase: ${this.apiBase},
        loginUrl: ${this.loginUrl}
      `);
      console.error('The [siteId], [slug], [apiBase] and [loginUrl] is required!');
      return;
    }

    this.talkee = new TalkeeSDK({
      // required
      siteId: this.siteId, // site id. 1 is debug env
      slug: this.slug, // post slug. the unique id to identify posts
      commentSelector: '#talkee', // selector for comments. should be a string or a dom object

      // optional
      prefixCls: this.prefixCls,
      expandable: this.expandable,
      tweetTags: this.tweetTags, // tweet tags
      // identitySelector: '#links-login',     // selector for identity.
      apiBase: this.apiBase,                     // alternative apiBase.
      loginUrl: this.loginUrl,                   // alternative login url.
      render: this.renderOpts,
      onError: this.handleError,
      redirectHash: false,
      hidePaginationWhenSinglePage: true
    });

    this.$emit('init', this.talkee);
  }

  public mounted() {
    this.initTalkee();
    this.resizeListener();
  }

  public beforeDestroy() {
    if (this.timer !== void 0) {
      clearTimeout(this.timer);
      this.timer = void 0;
    }
  }

  public render(h: CreateElement): VNode {
    return h(
      'div',
      {
        staticClass: this.classes('wrapper'),
        domProps: {
          id: 'talkee'
        },
        ref: 'talkee'
      }
    );
  }

  @Watch('siteId')
  @Watch('slug')
  @Watch('apiBase')
  @Watch('loginUrl')
  protected updateTalkee() {
    if (this.talkee) {
      this.talkee.setOptions({
        siteId: this.siteId,
        slug: this.slug,
        apiBase: this.apiBase,
        loginUrl: this.loginUrl
      });
      this.$emit('update', this.talkee);
    } else {
      this.initTalkee();
    }
  }
}

export default Talkee;