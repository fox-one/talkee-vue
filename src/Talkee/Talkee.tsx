import {
  ref,
  defineComponent,
  onBeforeMount,
  onMounted
} from '@vue/composition-api';
import classnames from '@utils/classnames';
import helper from '@utils/helper';
import { logger } from '@utils/logger';
import apis from '@apis/index';
import Comments from '../Comments';
import Editor from '../Editor';
import SortBar from '../SortBar';
import LoginBtn from '../LoginBtn';
/* import types */
import type { CreateElement, VNode } from 'vue';
import type { IComment } from '@/types/api';

export default defineComponent({
  name: 'Talkee',
  components: {
    Comments,
    Editor,
    SortBar,
    LoginBtn
  },
  props: {
    prefixCls: {
      type: String,
      default: 'talkee'
    },
    siteId: {
      type: [Number, String],
      default: null
    },
    slug: {
      type: String,
      default: null
    },
    apiBase: {
      type: String,
      default: null
    },
    loginUrl: {
      type: String,
      default: null
    },
    commentHeight: {
      type: String,
      default: '50vh'
    },
    commentLength: {
      type: [Number, String],
      default: 512
    },
    nestedId: {
      type: [Number, String],
      default: ''
    }
  },
  setup(props, context) {
    const {
      prefixCls,
      siteId,
      slug,
      apiBase,
      loginUrl
    } = props;
    const classes = classnames(prefixCls);
    const total = ref(0);
    const order = ref('favor_count');
    const isLogin = ref(!!(helper.getToken() && helper.getProfile()));

    onBeforeMount(async () => {
      if (siteId == null || slug == null || apiBase == null || loginUrl == null) {
        context.emit('error', 'missing params!');
        logger.error('missing params!');
        return;
      }
      helper.setDefaultParams({
        site_id: siteId,
        slug,
        api_base: apiBase,
        login_url: loginUrl
      });
      if (!isLogin.value) {
        const query: any = helper.getUrlQuery();
        if (query.code) {
          try {
            const auth = await apis.auth(query.code);
            helper.setAuth(auth);
            const me = await apis.getMe();
            helper.setProfile(me);
            isLogin.value = true;
            context.emit('login:success');
          } catch (err) {
            context.emit('login:fail', err);
            context.emit('error', err);
          }
        } else {
          helper.setRedirect();
        }
      }
    });

    onMounted(() => {
      context.emit('init', { ...props, isLogin: isLogin.value });
    });

    return {
      classes,
      total,
      order,
      isLogin
    };
  },
  methods: {
    handleKeyboard(state: 'rise' | 'fold') {
      this.$emit('keyboard', state);
    },
    handleError(e) {
      logger.error(e);
      this.$emit('error', e);
    },
    handleComment(comment: string) {
      this.total++;
      (this.$refs.comments as any)?.comments?.unshift?.(comment);
      (this.$refs.comments as any)?.scrollToTop?.();
      this.$emit('comment', comment);
    },
    handleSubComment(comment: string) {
      this.$emit('subcomment', comment);
    },
    handleOrderChange(type: string) {
      switch(type) {
        case 'asc':
          this.order = 'id-asc';
          break;
        case 'desc':
          this.order = 'id';
          break;
        default:
          this.order = 'favor_count';
          break;
      }
    }
  },
  render(h: CreateElement): VNode {
    const { prefixCls } = this.$props;

    return (
      <div class={this.classes('', 'd-flex flex-column')}>
        <SortBar
          vOn:error={this.handleError}
          vOn:sort={this.handleOrderChange}
          prefixCls={prefixCls}
          total={this.total}
          {...{ attrs: this.$attrs }}
        />
        <div class={this.classes('middle', 'mt-6 text-center')}>
          { this.isLogin ? <Editor
            vOn:keyboard={this.handleKeyboard}
            vOn:error={this.handleError}
            vOn:comment={this.handleComment}
            prefixCls={prefixCls}
            isLogin={this.isLogin}
            maxLength={this.commentLength}
            {...{ attrs: this.$attrs }}
          /> : <LoginBtn
            vOn:error={this.handleError}
            prefixCls={prefixCls}
            {...{ attrs: this.$attrs }}
          />}
        </div>
        <Comments
          vOn:keyboard={this.handleKeyboard}
          vOn:error={this.handleError}
          vOn:loaded={() => {
            this.total = (this.$refs.comments as any)?.total;
          }}
          vOn:subcomment={this.handleSubComment}
          order={this.order}
          height={this.commentHeight}
          maxLength={this.commentLength}
          nestedId={this.nestedId}
          isLogin={this.isLogin}
          class="pt-4"
          prefixCls={prefixCls}
          ref="comments"
          {...{ attrs: this.$attrs }}
        />
      </div>
    );
  }
});