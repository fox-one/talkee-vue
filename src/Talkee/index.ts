import TalkeeNext from './Talkee';
import type { defineComponent } from '@vue/composition-api';
import type { VueConstructor } from 'vue';

export type TalkeeNext = ReturnType<typeof defineComponent> & {
  install(app: VueConstructor<Vue>): void
}

(TalkeeNext as TalkeeNext).install = (app: VueConstructor<Vue>): void => {
  app.component(TalkeeNext.name, TalkeeNext);
};

export default TalkeeNext;
