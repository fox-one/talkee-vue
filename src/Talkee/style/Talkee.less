.prefix-talkee {
  display: block;
  color: green;
}
