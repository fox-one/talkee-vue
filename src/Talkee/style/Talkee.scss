.talkee {
  display: block;
  background-color: var(--v-greyscale_7-base);
}
