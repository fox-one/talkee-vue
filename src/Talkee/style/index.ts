import './talkee.less';
import './talkee-theme-group.less';