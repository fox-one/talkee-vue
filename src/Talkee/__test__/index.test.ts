import { shallowMount, mount } from '@vue/test-utils';
import Talkee from '../';

describe('Talkee', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(Talkee);
    expect(wrapper).toMatchSnapshot();
  });
});
