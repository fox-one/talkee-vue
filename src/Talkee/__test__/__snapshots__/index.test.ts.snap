// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Talkee renders a div 1`] = `
VueWrapper {
  "_emitted": Object {
    "init": Array [
      Array [
        Talkee {
          "_isLogin": false,
          "apiBase": "",
          "applySortMethod": [Function],
          "auth": [Function],
          "buildCommentUI": [Function],
          "buildEditorUI": [Function],
          "buildLoadingUI": [Function],
          "buildLoginURL": [Function],
          "buildTalkeeUI": [Function],
          "checkTextareaStatus": [Function],
          "components": Object {},
          "container": null,
          "defaultAvatarUrl": "",
          "editorArea": null,
          "expandable": false,
          "init": [Function],
          "isSigned": false,
          "itemPerPage": 10,
          "loginUrl": "",
          "opts": Object {
            "apiBase": "https://talkee-test-api.firesbox.com/api",
            "commentSelector": "#talkee",
            "expandable": true,
            "hidePaginationWhenSinglePage": true,
            "loginRedirect": false,
            "loginUrl": "https://mixin-oauth.firesbox.com/?client_id=65fdd5af-5d3f-496a-9d62-b5cdb4d8396f&scope=PROFILE:READ+PHONE:READ&response_type=code&redirect_url=",
            "prefixCls": "talkee",
            "render": undefined,
            "siteId": 6,
            "slug": "1234",
            "tweetTags": Array [],
          },
          "page": 1,
          "prefixCls": "talkee",
          "profile": null,
          "repliedCommentId": null,
          "repliedUserId": null,
          "sendComment": [Function],
          "siteId": "",
          "slug": "",
          "sortMethod": "favor_count",
          "total": 0,
          "totalPage": 1,
          "tweetTags": Array [],
        },
      ],
    ],
  },
  "_emittedByOrder": Array [
    Object {
      "args": Array [
        Talkee {
          "_isLogin": false,
          "apiBase": "",
          "applySortMethod": [Function],
          "auth": [Function],
          "buildCommentUI": [Function],
          "buildEditorUI": [Function],
          "buildLoadingUI": [Function],
          "buildLoginURL": [Function],
          "buildTalkeeUI": [Function],
          "checkTextareaStatus": [Function],
          "components": Object {},
          "container": null,
          "defaultAvatarUrl": "",
          "editorArea": null,
          "expandable": false,
          "init": [Function],
          "isSigned": false,
          "itemPerPage": 10,
          "loginUrl": "",
          "opts": Object {
            "apiBase": "https://talkee-test-api.firesbox.com/api",
            "commentSelector": "#talkee",
            "expandable": true,
            "hidePaginationWhenSinglePage": true,
            "loginRedirect": false,
            "loginUrl": "https://mixin-oauth.firesbox.com/?client_id=65fdd5af-5d3f-496a-9d62-b5cdb4d8396f&scope=PROFILE:READ+PHONE:READ&response_type=code&redirect_url=",
            "prefixCls": "talkee",
            "render": undefined,
            "siteId": 6,
            "slug": "1234",
            "tweetTags": Array [],
          },
          "page": 1,
          "prefixCls": "talkee",
          "profile": null,
          "repliedCommentId": null,
          "repliedUserId": null,
          "sendComment": [Function],
          "siteId": "",
          "slug": "",
          "sortMethod": "favor_count",
          "total": 0,
          "totalPage": 1,
          "tweetTags": Array [],
        },
      ],
      "name": "init",
    },
  ],
  "isFunctionalComponent": undefined,
}
`;

exports[`Talkee renders correctly 1`] = `
VueWrapper {
  "_emitted": Object {
    "init": Array [
      Array [
        Talkee {
          "_isLogin": false,
          "apiBase": "",
          "applySortMethod": [Function],
          "auth": [Function],
          "buildCommentUI": [Function],
          "buildEditorUI": [Function],
          "buildLoadingUI": [Function],
          "buildLoginURL": [Function],
          "buildTalkeeUI": [Function],
          "checkTextareaStatus": [Function],
          "components": Object {},
          "container": null,
          "defaultAvatarUrl": "",
          "editorArea": null,
          "expandable": false,
          "init": [Function],
          "isSigned": false,
          "itemPerPage": 10,
          "loginUrl": "",
          "opts": Object {
            "apiBase": "https://talkee-test-api.firesbox.com/api",
            "commentSelector": "#talkee",
            "expandable": true,
            "hidePaginationWhenSinglePage": true,
            "loginRedirect": false,
            "loginUrl": "https://mixin-oauth.firesbox.com/?client_id=65fdd5af-5d3f-496a-9d62-b5cdb4d8396f&scope=PROFILE:READ+PHONE:READ&response_type=code&redirect_url=",
            "prefixCls": "talkee",
            "render": undefined,
            "siteId": 6,
            "slug": "1234",
            "tweetTags": Array [],
          },
          "page": 1,
          "prefixCls": "talkee",
          "profile": null,
          "repliedCommentId": null,
          "repliedUserId": null,
          "sendComment": [Function],
          "siteId": "",
          "slug": "",
          "sortMethod": "favor_count",
          "total": 0,
          "totalPage": 1,
          "tweetTags": Array [],
        },
      ],
    ],
  },
  "_emittedByOrder": Array [
    Object {
      "args": Array [
        Talkee {
          "_isLogin": false,
          "apiBase": "",
          "applySortMethod": [Function],
          "auth": [Function],
          "buildCommentUI": [Function],
          "buildEditorUI": [Function],
          "buildLoadingUI": [Function],
          "buildLoginURL": [Function],
          "buildTalkeeUI": [Function],
          "checkTextareaStatus": [Function],
          "components": Object {},
          "container": null,
          "defaultAvatarUrl": "",
          "editorArea": null,
          "expandable": false,
          "init": [Function],
          "isSigned": false,
          "itemPerPage": 10,
          "loginUrl": "",
          "opts": Object {
            "apiBase": "https://talkee-test-api.firesbox.com/api",
            "commentSelector": "#talkee",
            "expandable": true,
            "hidePaginationWhenSinglePage": true,
            "loginRedirect": false,
            "loginUrl": "https://mixin-oauth.firesbox.com/?client_id=65fdd5af-5d3f-496a-9d62-b5cdb4d8396f&scope=PROFILE:READ+PHONE:READ&response_type=code&redirect_url=",
            "prefixCls": "talkee",
            "render": undefined,
            "siteId": 6,
            "slug": "1234",
            "tweetTags": Array [],
          },
          "page": 1,
          "prefixCls": "talkee",
          "profile": null,
          "repliedCommentId": null,
          "repliedUserId": null,
          "sendComment": [Function],
          "siteId": "",
          "slug": "",
          "sortMethod": "favor_count",
          "total": 0,
          "totalPage": 1,
          "tweetTags": Array [],
        },
      ],
      "name": "init",
    },
  ],
  "isFunctionalComponent": undefined,
}
`;
